cirq-core/cirq/ops/kraus_channel.py

```diff
@@ -0,0 +1,95 @@
+from typing import Any, Dict, Iterable, Optional, Tuple
+import numpy as np
+
+from cirq import protocols, value
+from cirq._compat import proper_repr
+from cirq.ops import raw_types
+
+
+class KrausChannel(raw_types.Gate):
+    """A generic channel that can record the index of its selected operator.
+
+    Args:
+        kraus_ops: a list of Kraus operators, formatted as numpy array.
+        key: an optional measurement key string for this channel. Simulations
+            which select a single Kraus operator to apply will store the index
+            of that operator in the measurement result list with this key.
+    """
+
+    def __init__(self, kraus_ops: Iterable[np.ndarray], key: Optional[str] = None):
+        # TODO: validate channel representations (issue #2271)
+        kraus_ops = list(kraus_ops)
+        if not kraus_ops:
+            raise ValueError('KrausChannel must have at least one operation.')
+        num_qubits = np.log2(kraus_ops[0].size) / 2
+        if not num_qubits.is_integer():
+            raise ValueError(
+                f'Input Kraus ops of shape {kraus_ops[0].shape} does not '
+                'represent an operator over qubits.'
+            )
+        self._num_qubits = int(num_qubits)
+        for i, op in enumerate(kraus_ops):
+            if not op.size == kraus_ops[0].size:
+                raise ValueError(
+                    'Inconsistent Kraus operator sizes: '
+                    f'op[0]: {kraus_ops[0].size}, op[{i}]: {op.size}'
+                )
```

Comment:
Do we care if these ops are CPTP ?

![image](https://user-images.githubusercontent.com/2791105/127388967-bea9b427-febc-4ad9-9d61-3638fdca3e45.png)

where `op` is the list of kraus_ops.

Reply:
Added option to validate. Defaults to false due to the cost of this check.

Reply:
Note that the cost of this check is the same as the cost of evaluating the channel on a single input (see below for some interesting details). If this cost is too high, then perhaps we need to speed things up elsewhere. If it isn't then I'd say adding the `validate` argument may be premature optimization and we should wait until someone complains about instantiation cost.

Here is the current timing:

```
In [5]: %timeit cirq.is_cptp(kraus_ops=[np.eye(2)/2, cirq.unitary(cirq.X)/2, cirq.unitary(cirq.Y)/2, cirq.unitary(cirq.Z)/2])
212 µs ± 3.26 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

so for a single-qubit channel with maximum number of Kraus operators we can do ~5,000 checks per second. Since this is python, I'd say it's probably acceptable.

----
For every channel E, there is an adjoint channel E^\dagger. If K_1,..., K_n are Kraus operators of E, then K_1^\dagger, ..., K_n^\dagger are Kraus operators of E^\dagger. Also, one can prove that E preserves the trace if and only if E^\dagger is unital, i.e. if E^\dagger maps identity matrix to identity matrix. Substituting into the formula for E^\dagger in terms of its Kraus operators yields the identity that Michael posted above. Adjoint channels have a [nice interpretation](https://quantumcomputing.stackexchange.com/a/12927/10480).

Reply:
There is an open issue (#408) for defining `_kraus_` (previously `_channel_`) on `cirq.Circuit` objects. Such an operation would need to construct a combinatorially-exploding number of Kraus operators, potentially over many qubits, and would not require validation if the input channels were known to be CPTP maps. I see this as a strong motivation to keep the `validate` argument.

> If this cost is too high, then perhaps we need to speed things up elsewhere.

Do you have recommendations for this? Should these optimizations be part of this PR?

Reply:
Thanks for the pointer to #408. I needed this in the past and wrote the code that solve the problem. I'll clean it up, add tests and will send it out (probably next week). The solution avoids combinatorial explosion of Kraus operators entirely by using a different representation (see #4179), so I'm not concerned about validation cost in this context.

I don't have a recommendation for speeding up `cirq.is_cptp`. It looks pretty good already.

After second thought, I think there is really no harm in the extra control over whether validation happens, so feel free to keep it. That said, I do think we should make the default `validate=True`.